src/ngx_http_lua_socket_tcp.c

```diff
@@ -6395,4 +6392,142 @@ ngx_http_lua_ffi_socket_tcp_del_udata(ngx_http_lua_socket_tcp_upstream_t *u,
+    int option, int *val, char **err_msg)
+{
+    socklen_t len;
+    int       fd;
+
+    if (u == NULL || u->peer.connection == NULL) {
+        *err_msg = "closed";
+        return NGX_ERROR;
+    }
+
+    fd = u->peer.connection->fd;
+    len = sizeof(int);
+    switch (option) {
+    case OPTION_NAME_KEEPALIVE:
+        if (getsockopt(fd, SOL_SOCKET, SO_KEEPALIVE, (void *) val, &len)
+            == -1)
+        {
+            *err_msg = "set keepalive failed";
+            return NGX_ERROR;
+        }
+        break;
+
+    case OPTION_NAME_REUSEADDR:
+        if (getsockopt(fd, SOL_SOCKET, SO_REUSEADDR, (void *) val, &len)
+            == -1)
+        {
+            *err_msg = "set reuseaddr failed";
+            return NGX_ERROR;
+        }
+        break;
+
+    case OPTION_NAME_TCP_NODELAY:
+        if (getsockopt(fd, IPPROTO_TCP, TCP_NODELAY, (void *) val, &len)
+            == -1)
+        {
+            *err_msg = "set tcp-nodelay failed";
+            return NGX_ERROR;
+        }
+        break;
+
+    case OPTION_NAME_SNDBUF:
+        if (getsockopt(fd, SOL_SOCKET, SO_RCVBUF, (void *) val, &len)
+            == -1)
+        {
+            *err_msg = "set rcvbuf failed";
+            return NGX_ERROR;
+        }
+        break;
+
+    case OPTION_NAME_RCVBUF:
+        if (getsockopt(fd, SOL_SOCKET, SO_SNDBUF, (void *) val, &len)
+            == -1)
+        {
+            *err_msg = "set sndbuf failed";
+            return NGX_ERROR;
+        }
+        break;
+
+    default:
+        *err_msg = "unsupported option";
+        return NGX_ERROR;
+    }
+
+    return NGX_OK;
+}
+
+
+int
+ngx_http_lua_ffi_socket_tcp_setoption(ngx_http_lua_socket_tcp_upstream_t *u,
+    int option, int val, char **err_msg)
+{
+    socklen_t len;
+    int       fd;
+
+    if (u == NULL || u->peer.connection == NULL) {
+        *err_msg = "closed";
+        return NGX_ERROR;
+    }
+
+    len = sizeof(int);
+    fd = u->peer.connection->fd;
+    switch (option) {
+    case OPTION_NAME_KEEPALIVE:
+        if (setsockopt(fd, SOL_SOCKET, SO_KEEPALIVE, (const void *) &val, len)
+            == -1)
+        {
+            *err_msg = "set keepalive failed";
+            return NGX_ERROR;
+        }
+        break;
+
+    case OPTION_NAME_REUSEADDR:
+        if (setsockopt(fd, SOL_SOCKET, SO_REUSEADDR, (const void *) &val, len)
+            == -1)
+        {
+            *err_msg = "set reuseaddr failed";
+            return NGX_ERROR;
+        }
+        break;
+
+    case OPTION_NAME_TCP_NODELAY:
+        if (setsockopt(fd, IPPROTO_TCP, TCP_NODELAY, (const void *) &val, len)
+            == -1)
+        {
+            *err_msg = "set tcp-nodelay failed";
+            return NGX_ERROR;
+        }
+        break;
+
+    case OPTION_NAME_SNDBUF:
+        if (setsockopt(fd, SOL_SOCKET, SO_RCVBUF, (const void *) &val, len)
+            == -1)
+        {
+            *err_msg = "set rcvbuf failed";
```

Comment:
We should return the system error message as well, otherwise it's hard to debug (if not impossible). This applies to the other setsockopt calls in this PR.

Reply:
accept